File: p1.py
#Time; O(N)
#Space: O(1)
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        res=0
        n=len(height)

        lw=height[0]
        rw=height[n-1]

        l=0
        r=n-1

        while(l<=r):
            if(lw<=rw):
                if(height[l]<lw):
                    res+=lw-height[l]
                else:
                    lw=height[l]
                l+=1
            else:
                if(height[r]<rw):
                    res+=rw-height[r]
                else:
                    rw=height[r]
                r-=1
        return res

File: p1.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach



class Solution {
    public int trap(int[] height) {
        int temp = 0;
        int l = height.length;
        int[] left = new int[l];
        int[] right = new int[l];

        left[0] = 0;
        right[l-1]=0;
        int j=l-1;
        //Get the max left wall and right wall for current
        for(int i =1; i<l; i++){
            left[i] = Math.max(left[i-1],height[i-1]);
            right[j-i] = Math.max(height[j-i+1],right[j-i+1]);
        }
        int temp2 = 0;
        //get the minimun out of left and right wall to calculate how much water will be stored
        for(int i = 0; i<l; i++){
            temp2 = Math.min(left[i], right[i]) - height[i];
                if(temp2>0){
                    temp = temp + temp2;
                }
            
        }

        return temp;
    }
}